# Análise de Dados com as Bases da Empresa Contoso

## Importação de Bibliotecas e Arquivos

In [1]:
# Importa a biblioteca pandas com o apelido pd
import pandas as pd

In [2]:
# Importação das bases de dados
df_vendas = pd.read_csv("Contoso - Vendas - 2017.csv", sep=";")
df_produtos = pd.read_csv("Contoso - Cadastro Produtos.csv", sep=";")
df_lojas = pd.read_csv("Contoso - Lojas.csv", sep=";")
df_clientes = pd.read_csv("Contoso - Clientes.csv", sep=";")

## Visualização dos Dataframes

### Vendas

In [3]:
df_vendas.head()

Numero da Venda Data da Venda Data do Envio  ID Canal  ID Loja  ID Produto  \
0                1    01/01/2017    02/01/2017         1       86         981   
1                2    01/01/2017    06/01/2017         5      308        1586   
2                3    01/01/2017    01/01/2017         0      294        1444   
3                4    01/01/2017    01/01/2017         0      251        1468   
4                5    01/01/2017    07/01/2017         6       94        1106   

   ID Promocao  ID Cliente  Quantidade Vendida  Quantidade Devolvida  
0            2        6825                   9                     1  
1            2       18469                   9                     1  
2            5       19730                  13                     1  
3            5       29326                   6                     1  
4            2       22617                   4                     1

### Produtos

In [4]:
df_produtos.head()

Nome do Produto  \
0            Contoso Wireless Laser Mouse E50 Grey   
1    Contoso Optical Wheel OEM PS/2 Mouse E60 Grey   
2   Contoso Optical Wheel OEM PS/2 Mouse E60 Black   
3   Contoso Optical Wheel OEM PS/2 Mouse E60 White   
4  Contoso Optical Wheel OEM PS/2 Mouse E60 Silver   

                                Descricao do Produto    Fabricante  \
0  Advanced 2.4 GHz cordless technology makes fre...  Contoso, Ltd   
1                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
2                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
3                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
4                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   

  Nome da Marca       Tipo Custo Unitario Preco Unitario  ID Produto  \
0       Contoso  Econômico          10,69          20,96         873   
1       Contoso  Econômico           6,63             13         879   
2       Contoso  Econômico           6,63             13         880   
3       Contoso  Econômico           6,63             13         881   
4       Contoso  Econômico           6,63             13         882   

   ID Subcategoria  
0               22  
1               22  
2               22  
3               22  
4               22

### Lojas

In [5]:
df_lojas.head()

ID Loja                Nome da Loja  Quantidade Colaboradores  \
0        1  Loja Contoso Seattle No.1                       17.0   
1        2  Loja Contoso Seattle No.2                       25.0   
2        3     Loja Contoso Kennewick                       26.0   
3        4      Loja Contoso Bellevue                       19.0   
4        5       Loja Contoso Redmond                       33.0   

             País  
0  Estados Unidos  
1  Estados Unidos  
2  Estados Unidos  
3  Estados Unidos  
4  Estados Unidos

### Clientes

In [6]:
df_clientes.head()

ID Cliente Primeiro Nome     Sobrenome                       E-mail  \
0           1        Garnet    Lanfranchi  glanfranchi0@mayoclinic.com   
1           2       Lurette     Roseblade    lroseblade1@bigcartel.com   
2           3       Glenden     Ishchenko    gishchenko2@moonfruit.com   
3           4         Baron  Jedrzejewsky  bjedrzejewsky3@e-recht24.de   
4           5        Sheree      Bredbury     sbredbury4@sitemeter.com   

      Genero  Numero de Filhos Data de Nascimento  Unnamed: 7  Unnamed: 8  \
0   Feminino                 2         12/05/1995         NaN         NaN   
1   Feminino                 2         30/06/1943         NaN         NaN   
2  Masculino                 5         09/04/1989         NaN         NaN   
3  Masculino                 4         17/11/1998         NaN         NaN   
4   Feminino                 5         08/09/1975         NaN         NaN   

   Unnamed: 9  Unnamed: 10  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN

## Limpeza dos Dados

Repare que o `df_clientes` possui quatro colunas com dados vazios. Vamos, pois, remover essas colunas.

In [7]:
df_clientes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39504 entries, 0 to 39503
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID Cliente          39504 non-null  int64  
 1   Primeiro Nome       39504 non-null  object 
 2   Sobrenome           39504 non-null  object 
 3   E-mail              39504 non-null  object 
 4   Genero              39504 non-null  object 
 5   Numero de Filhos    39504 non-null  int64  
 6   Data de Nascimento  39504 non-null  object 
 7   Unnamed: 7          0 non-null      float64
 8   Unnamed: 8          0 non-null      float64
 9   Unnamed: 9          0 non-null      float64
 10  Unnamed: 10         0 non-null      float64
dtypes: float64(4), int64(2), object(5)
memory usage: 3.3+ MB


In [8]:
# Remove as colunas especificadas
df_clientes.drop(["Unnamed: 7", "Unnamed: 8", "Unnamed: 9", "Unnamed: 10"], axis=1, inplace=True)
df_clientes.head()

ID Cliente Primeiro Nome     Sobrenome                       E-mail  \
0           1        Garnet    Lanfranchi  glanfranchi0@mayoclinic.com   
1           2       Lurette     Roseblade    lroseblade1@bigcartel.com   
2           3       Glenden     Ishchenko    gishchenko2@moonfruit.com   
3           4         Baron  Jedrzejewsky  bjedrzejewsky3@e-recht24.de   
4           5        Sheree      Bredbury     sbredbury4@sitemeter.com   

      Genero  Numero de Filhos Data de Nascimento  
0   Feminino                 2         12/05/1995  
1   Feminino                 2         30/06/1943  
2  Masculino                 5         09/04/1989  
3  Masculino                 4         17/11/1998  
4   Feminino                 5         08/09/1975

Neste ponto, vamos selecionar apenas as colunas que nos interessam:

In [9]:
# Seleciona as colunas especificadas
df_clientes = df_clientes[["ID Cliente", "Primeiro Nome", "Sobrenome", "E-mail"]]
df_produtos = df_produtos[["ID Produto", "Nome do Produto"]]
df_lojas = df_lojas[["ID Loja", "Nome da Loja"]]

Agora, vamos criar um único dataframe, lembrando que o `df_vendas` é uma tabela fato e que os demais são tabelas característica:

In [10]:
# Mescla os dataframes com base em uma coluna em comum
df_vendas = df_vendas.merge(df_clientes, on="ID Cliente")
df_vendas = df_vendas.merge(df_produtos, on="ID Produto")
df_vendas = df_vendas.merge(df_lojas, on="ID Loja")

Vamos renomear a coluna __E-mail__ para __E-mail do Cliente__:

In [11]:
# Renomeia uma coluna do dataframe
df_vendas = df_vendas.rename(columns={"E-mail": "E-mail do Cliente"})
df_vendas.head()

Numero da Venda Data da Venda Data do Envio  ID Canal  ID Loja  ID Produto  \
0                1    01/01/2017    02/01/2017         1       86         981   
1           372597    21/05/2017    22/05/2017         1       86         981   
2           373159    22/05/2017    24/05/2017         2       86         448   
3           338147    10/05/2017    10/05/2017         0       86         448   
4           368615    20/05/2017    26/05/2017         6       86         448   

   ID Promocao  ID Cliente  Quantidade Vendida  Quantidade Devolvida  \
0            2        6825                   9                     1   
1            1       21344                  10                     0   
2            1       19328                  15                     1   
3            1       12792                  15                     0   
4            1       19820                  15                     0   

  Primeiro Nome  Sobrenome             E-mail do Cliente  \
0         Rurik  Brumfield        rbrumfieldmy@ameblo.jp   
1       Makayla    Sanders  makayla3@adventure-works.com   
2         Jolee    Gundrey        jgundrey97@youtube.com   
3         Lucio  Tomlinson          ltomlinsonm1@gnu.org   
4           Pen  Rudeforth  prudeforthmv@themeforest.net   

                              Nome do Produto          Nome da Loja  
0  A. Datum Advanced Digital Camera M300 Pink  Loja Contoso Austin   
1  A. Datum Advanced Digital Camera M300 Pink  Loja Contoso Austin   
2              WWI Desktop PC1.80 E1801 Black  Loja Contoso Austin   
3              WWI Desktop PC1.80 E1801 Black  Loja Contoso Austin   
4              WWI Desktop PC1.80 E1801 Black  Loja Contoso Austin

## Análises

Responderemos as seguintes perguntas:

1. Qual o cliente que comprou mais vezes?

2. Qual a loja que mais vendeu?

3. Qual o produto que menos vendeu?

4. Qual o percentual de vendas que foram devolvidas?

5. Qual foi o percentual de devoluções da Loja Contoso Europe Online?

6. Quantas vendas da Loja Contoso Europe Online não tiveram devoluções?

### Cliente que mais comprou

In [12]:
pedidos_clientes = df_vendas["ID Cliente"].value_counts()
pedidos_clientes

21485    56
16984    56
20009    56
26090    53
13433    53
         ..
39300     1
732       1
379       1
991       1
815       1
Name: ID Cliente, Length: 38930, dtype: int64

Assim, o cliente que comprou mais vezes foi o de ID 21485, tendo efetuado 56 transações.

### Loja que mais vendeu

In [13]:
df_vendas_lojas = df_vendas[["Nome da Loja", "Quantidade Vendida"]]
df_vendas_lojas.head()

Nome da Loja  Quantidade Vendida
0  Loja Contoso Austin                    9
1  Loja Contoso Austin                   10
2  Loja Contoso Austin                   15
3  Loja Contoso Austin                   15
4  Loja Contoso Austin                   15

In [14]:
df_vendas_lojas_agrupadas = df_vendas_lojas.groupby("Nome da Loja").sum()
df_vendas_lojas_agrupadas.head()

Quantidade Vendida
Nome da Loja                                
Loja Contoso Albany                    26353
Loja Contoso Alexandria                26247
Loja Contoso Amsterdam                 28294
Loja Contoso Anchorage                 27451
Loja Contoso Annapolis                 26065

In [15]:
# Maior valor
maior_valor = df_vendas_lojas_agrupadas["Quantidade Vendida"].max()

# Melhor loja
melhor_loja = df_vendas_lojas_agrupadas["Quantidade Vendida"].idxmax()

# Exibição de resultado
print(f"A loja que mais vendeu foi a {melhor_loja}, com {maior_valor} unidades vendidas.")

A loja que mais vendeu foi a Loja Contoso Catalog , com 1029117 unidades vendidas.


### Produto que menos vendeu

In [16]:
df_vendas_produtos = df_vendas[["Nome do Produto", "Quantidade Vendida"]]
df_vendas_produtos.head()

Nome do Produto  Quantidade Vendida
0  A. Datum Advanced Digital Camera M300 Pink                   9
1  A. Datum Advanced Digital Camera M300 Pink                  10
2              WWI Desktop PC1.80 E1801 Black                  15
3              WWI Desktop PC1.80 E1801 Black                  15
4              WWI Desktop PC1.80 E1801 Black                  15

In [17]:
df_vendas_produtos_agrupadas = df_vendas_produtos.groupby("Nome do Produto").sum()
df_vendas_produtos_agrupadas.head()

Quantidade Vendida
Nome do Produto                                                 
A. Datum Advanced Digital Camera M300 Azure                 8999
A. Datum Advanced Digital Camera M300 Black                12087
A. Datum Advanced Digital Camera M300 Green                11400
A. Datum Advanced Digital Camera M300 Grey                 10713
A. Datum Advanced Digital Camera M300 Orange               12181

In [18]:
# Maior quantidade vendida
maior_qtde_vendida = df_vendas_produtos_agrupadas["Quantidade Vendida"].max()

# Produto mais vendido
produto_mais_vendido = df_vendas_produtos_agrupadas["Quantidade Vendida"].idxmax()

# Exibição de resultados
print(f"O produto mais vendido foi o {produto_mais_vendido}, com {maior_qtde_vendida} unidades vendidas.")

O produto mais vendido foi o Cigarette Lighter Adapter for Contoso Phones E110 White, com 114300 unidades vendidas.


### Percentual de pedidos que foram devolvidos

In [19]:
# Cálculos
qtde_pedidos = df_vendas["Quantidade Vendida"].sum()
qtde_devolvida = df_vendas["Quantidade Devolvida"].sum()
percentual_devolucoes = qtde_devolvida / qtde_pedidos

# Exibição de resultado
print(f"Percentual de devoluções: {percentual_devolucoes:.2%}")

Percentual de devoluções: 1.17%


### Percentual de devoluções da Contoso Europe Online

O ID da Loja Contoso Europe Online é 306. Vamos usá-lo:

In [20]:
df_vendas_contoso_europe_online = df_vendas.query("`ID Loja` == 306")
df_vendas_contoso_europe_online.head()

Numero da Venda Data da Venda Data do Envio  ID Canal  ID Loja  \
322802             2007    01/01/2017    02/01/2017         1      306   
322803            52480    22/01/2017    22/01/2017         0      306   
322804           973865    29/12/2017    31/12/2017         2      306   
322805           160143    07/03/2017    10/03/2017         3      306   
322806           757619    12/10/2017    12/10/2017         0      306   

        ID Produto  ID Promocao  ID Cliente  Quantidade Vendida  \
322802         981           10       23908                  16   
322803         981           10       17100                   8   
322804         981           10       10714                   8   
322805         981            8       18707                  12   
322806         981           10       14169                   8   

        Quantidade Devolvida Primeiro Nome  Sobrenome  \
322802                     0         Lydia      Lopez   
322803                     0       Margaux  Spohrmann   
322804                     0        Sancho    Kenrack   
322805                     1         Hulda     Becken   
322806                     0          Yves   Mougenel   

                    E-mail do Cliente  \
322802    lydia15@adventure-works.com   
322803  mspohrmann2v@businessweek.com   
322804              skenrackjv@pen.io   
322805           hbeckenjq@sbwire.com   
322806          ymougenel4q@naver.com   

                                   Nome do Produto  \
322802  A. Datum Advanced Digital Camera M300 Pink   
322803  A. Datum Advanced Digital Camera M300 Pink   
322804  A. Datum Advanced Digital Camera M300 Pink   
322805  A. Datum Advanced Digital Camera M300 Pink   
322806  A. Datum Advanced Digital Camera M300 Pink   

                       Nome da Loja  
322802  Loja Contoso Europe Online   
322803  Loja Contoso Europe Online   
322804  Loja Contoso Europe Online   
322805  Loja Contoso Europe Online   
322806  Loja Contoso Europe Online

In [21]:
# Cálculos
pedidos_contoso_europe_online = df_vendas_contoso_europe_online["Quantidade Vendida"].sum()
devolucoes_contoso_europe_online = df_vendas_contoso_europe_online["Quantidade Devolvida"].sum()
percentual_devolucoes_contoso_europe_online = devolucoes_contoso_europe_online / pedidos_contoso_europe_online

# Exibição de resultados
print(f"Percentual de devoluções da Loja Contoso Europe Online: {percentual_devolucoes_contoso_europe_online:.2%}")

Percentual de devoluções da Loja Contoso Europe Online: 1.33%


### Número de vendas da Contoso Europe Online sem devoluções

In [22]:
# Dataframe filtrado
df_vendas_contoso_europe_online_sem_devolucoes = df_vendas.query("`ID Loja` == 306 and `Quantidade Devolvida` == 0")

# Número de vendas sem devoluções
vendas_sem_devolucao = df_vendas_contoso_europe_online_sem_devolucoes["Quantidade Vendida"].sum()

# Exibição de resultado
print(f"A Loja Contoso Europe Online teve {vendas_sem_devolucao} pedidos sem devoluções.")

A Loja Contoso Europe Online teve 481657 pedidos sem devoluções.
